demos/python_demos/formula_recognition_demo/utils.py

```diff
@@ -1,11 +1,340 @@
+        self.args = args
+        log.info("Creating Inference Engine")
+        self.ie = IECore()
+        self.ie.set_config(
+            {"PERF_COUNT": "YES" if self.args.perf_counts else "NO"}, args.device)
+        self.encoder = read_net(self.args.m_encoder, self.ie)
+        self.dec_step = read_net(self.args.m_decoder, self.ie)
+        self.exec_net_encoder = self.ie.load_network(network=self.encoder, device_name=self.args.device)
+        self.exec_net_decoder = self.ie.load_network(network=self.dec_step, device_name=self.args.device)
+        self.images_list = []
+        self.vocab = Vocab(self.args.vocab_path)
+        self.model_status = Model.Status.ready
+        self.is_async = args.interactive_mode
+        self.num_infers_decoder = 0
+        if not args.interactive_mode:
+            self.preprocess_inputs()
+
+    def preprocess_inputs(self):
+        batch_dim, channels, height, width = self.encoder.input_info['imgs'].input_data.shape
+        assert batch_dim == 1, "Demo only works with batch size 1."
+        assert channels in (1, 3), "Input image is not 1 or 3 channeled image."
+        target_shape = (height, width)
+        if os.path.isdir(self.args.input):
+            inputs = sorted(os.path.join(self.args.input, inp)
+                            for inp in os.listdir(self.args.input))
+        else:
+            inputs = [self.args.input]
+        log.info("Loading and preprocessing images")
+        for filenm in tqdm(inputs):
+            image_raw = cv.imread(filenm)
+            assert image_raw is not None, "Error reading image {}".format(filenm)
+            image = preprocess_image(
+                PREPROCESSING[self.args.preprocessing_type], image_raw, target_shape)
+            record = dict(img_name=filenm, img=image, formula=None)
+            self.images_list.append(record)
+
+    def _async_infer_encoder(self, image, req_id):
+        return self.exec_net_encoder.start_async(request_id=req_id, inputs={self.args.imgs_layer: image})
+
+    def _async_infer_decoder(self, row_enc_out, dec_st_c, dec_st_h, output, tgt, req_id):
+        self.num_infers_decoder += 1
+        return self.exec_net_decoder.start_async(request_id=req_id, inputs={self.args.row_enc_out_layer: row_enc_out,
+                                                                            self.args.dec_st_c_layer: dec_st_c,
+                                                                            self.args.dec_st_h_layer: dec_st_h,
+                                                                            self.args.output_prev_layer: output,
+                                                                            self.args.tgt_layer: tgt
+                                                                            }
+                                                 )
+
+    def infer_async(self, model_input):
+        model_input = change_layout(model_input)
+        assert self.is_async
+        # asynchronous variant
```

Comment:
This probably should be a docstring or completely removed as `infer_async` suggests that.

Reply:
Removed comment